woocommerce/pinterest-for-woocommerce
# Onboarding inbox notifications.

### Changes proposed in this Pull Request:

Add inbox notifications for various events related to the onboarding process:
- complete onboarding ( after 3, 7, 14, 30 days )
- enable sync
- sync error


### Screenshots:
Complete onboarding after 3 days:
![image](https://user-images.githubusercontent.com/17271089/161932905-661ba0d2-707a-4021-9b3c-72a2bb1074a1.png)

Complete onboarding after 7, 14, 30 days:
![image](https://user-images.githubusercontent.com/17271089/161937046-46e087bb-116a-4238-8a5e-3221c1b31b7b.png)

Enable Sync:
![image](https://user-images.githubusercontent.com/17271089/161939316-d76a76cb-e63f-41ad-9617-a50a9cab3dce.png)

Sync Error:
![image](https://user-images.githubusercontent.com/17271089/161945276-04dde361-0053-44a9-8bce-7178ebe03d7c.png)




### Detailed test instructions:


Since the inbox code is run once a day, via immediate action scheduler action that gets set up by cron heartbeat it is useful to have a way to trigger the cron actions. This can be done using a plugin like **WP Crontrol** or using **wpcli**:
```bash
wp cron event run pinterest_for_woocommerce_daily_heartbeat_cron
```

1. Complete onboarding after 3 days inbox message gets sent when the merchant has installed the plugin, but has not finished the onboarding and has 5 or more completed purchases:
    1. Install the plugin
    2. Do not complete the setup
    3. Have 5 completed orders.
    4. Trigger the daily heartbeat
    5. Wait 3 days ( kidding ), check wp_option  `pinterest-for-woocommerce-marketing-notifications-init-timestamp` and remove 4 days' worth of seconds from it.
    6. Trigger daily heartbeat again. 
    7. Go to WooCommerce > Home, and check the inbox. You may need to refresh the location once or twice before the message gets rendered.
    8. Click `Complete Setup` to check that the link works.


2. Complete onboarding after 7, 14, and 30 days. This is the same scenario as with 1. Just the timestamp option needs to be pulled back in time more.

3. Enable sync note is issued when the account is connected, and the product sync is disabled in settings.
     1. Connect account
     2. Go to Marketing > Pinterest > Settings and disable `Enable Product Sync`
     3. A new timestamp was recorded after connection. This time `pinterest-for-woocommerce-account-connection-timestamp`. Reduce 4 days' worth of time.
     4. Trigger daily heartbeat.
     5. Visit WooCommerce > Home to check that the Enable sync message is there.
     6. Check that the message link works.

4. Sync Error note is sent when the connection to the account is completed, sync is enabled but there are errors in the ingestion process. **If testing of this will not be feasible please review the code - I have tested this.**
    1. Setup the plugin and connect.
    2. Make the ingestion fail ( ¯\_(ツ)_/¯ ) for example leave the account connected for a day but make the feed not accessible ( disable local site ), so the next feed fetch will fail.
    3. When you see the error in the feed status trigger the daily hearbeat.
    4. Visit WooCommerce > Home to check that the Error sync message is there.

### Additional details:

- In order to minimize the resources footprint a daily heartbeat process was added. This allows testing the notes trigger condition just once a day. It may mean that the note's trigger time is offset +1 day but that does not change the usability of this solution. Making it more exact with timing would make the code much more complicated without a real benefit.

- When the plugin gets installed we don't have any information about when that has happened. We also do not have any information about when a successful connection has been made. Both of these values are required to know when a message should be sent to the inbox. For this purpose, this code adds two timestamp markers that help us to assume missing values.
The first marker is set when the inbox messages for the plugin code is triggered for the first time: https://github.com/woocommerce/pinterest-for-woocommerce/blob/09aa302659fc2c6662d1c860fa8be174c3a758ba/src/Notes/MarketingNotifications.php#L69 First call sets the marker. The notifications use this is a general start time ( fake plugin install time ). Because we can't know when the plugin was installed or not. This is only for the purpose of the Notifications so this is why we have it here. <br>
The second time is added to general utilities. The role of this timer is to catch the connection time. It is tied to the last step of the connection process. If the user changes tag settings this will get reset which makes sense because the connection parameters have been changed. ofc this will also get reset on a full reconnection. 
https://github.com/woocommerce/pinterest-for-woocommerce/blob/09aa302659fc2c6662d1c860fa8be174c3a758ba/src/Utilities/Utilities.php#L45


### Changelog entry

> Add - Inbox notifications for the onboarding process and catalog sync.
